tensorflow/python/training/saver.py

```diff
@@ -242,12 +242,19 @@ def _AddShardedSaveOpsForV2(self, checkpoint_prefix, per_device):
     #     <train dir>/
     #        myckpt{.index, .data-?????-of-?????}
     #
+    #   Filesystems with eventual consistency (such as S3), don't need a temporary
+    #   location. Using a temporary directory in those cases
+    #   might cause situations where files are not available during copy.
+    #
     # Users only need to interact with the user-specified prefix, which is
     # "<train dir>/myckpt" in this case.  Save() and Restore() work with the
     # prefix directly, instead of any physical pathname.  (On failure and
     # subsequent restore, an outdated and orphaned temporary directory can be
     # safely removed.)
-    _SHARDED_SUFFIX = "_temp_%s/part" % uuid.uuid4().hex
+    _SHARDED_SUFFIX = control_flow_ops.cond(
+        string_ops.regex_full_match(checkpoint_prefix, '^s3://.*'),
+        lambda: ".part",
+        lambda: "_temp_%s/part" % uuid.uuid4().hex)
```

Comment:
Would it be possible to use Python code in here instead of TF code?

Otherwise, this causes a change in graphdefs as now graphs suddenly need to also register that they use a `cond` op.

Reply:
One option is smart_cond: https://github.com/tensorflow/tensorflow/blob/7f0b31af816a4f9936b35c02544e383f1b8c090a/tensorflow/python/framework/smart_cond.py#L27

But I'm not sure that'll actually fix most GraphDef changes, since in 1.x the checkpoint prefix is a placeholder when writing a SaverDef, and we actually feed different values at runtime when constructing a new Saver. It's worth trying, but we'll probably need to fix the GraphDef tests even if we use smart_cond.

The other option would be to make something unconditional... my impression was that always changing the temporary directory structure in this way would not work, but maybe there's something we could do unconditionally?

Reply:
It wasn't possible to use python code because `checkpoint_prefix` is a tf graph construct and not a regular python string. 
Yeah as Allen said this depends on a placeholder. Is that not the case in TF2? 

Hmm making this unconditional would be possible if this temp file logic was in the file system layer. Trying to think of a way. Let me know if you think of one.

Reply:
I'll try arguing with the unit test or its owner. I don't think it should block import of your change.

Reply:
So now that I look closely, we're not actually running different operations on the cond branches. So tf.where works here, just switching between the two constants. I think that will be much less traumatic; unfortunately control flow is a bit special.

Without objection I'll just make that change on import rather than messing with the unit tests. There are some lint nits I ended up changing anyway. That also lets us keep the saved_model reference cycle test.

Reply:
I see. I was thinking about moving this change to the C++ op where this filename is available. It might be cleaner that way.

https://github.com/tensorflow/tensorflow/blob/1c7d512e38bcb76226e9fb767934ca8bd2a39002/tensorflow/core/kernels/save_op.cc#L63

Something like get the condition using `tensorflow::Env::Default()->HasAtomicMove()`
Then do some string magic to remove the temp and suffix. What do you think about that?

Reply:
It's fairly attractive to have the op actually write to the path you actually pass it. Otherwise we need a magic string that gets rewritten, which seems much less readable than having the condition in the graph itself.

I can cancel the commit if you want, though. Or that can be a followup? I think with s/cond/where/g this change will go in fine and not affect non-s3 filesystem users.

Reply:
Update: where + regexp ops mean we need to modify one team's selective kernel registrations. I'm negotiating with them, but if you want to push the logic down into the saving kernel that's fine too (I still think it's a somewhat unfortunate workaround).

Reply:
Should I wait for some time then before trying the change in the op itself? For now I've pushed a change from cond to where and verified its working.

Reply:
Yeah it's just that one selective registration issue and a couple people who have GraphDef goldens. I think we can get it in with the where change. Thanks.